File: Max Heap/MaxHeap.java
import java.util.Arrays;
import java.util.Scanner;

public class MaxHeap {
    int[] arr;
    int maxSize;
    int heapSize;

    MaxHeap(int maxSize) {
        this.maxSize = maxSize;
        arr = new int[maxSize];
        heapSize = 0;
    }

    void MaxHeapify(int i) {
        int l = lChild(i);
        int r = rChild(i);
        int largest = i;
        if (l < heapSize && arr[l] > arr[i])
            largest = l;
        if (r < heapSize && arr[r] > arr[largest])
            largest = r;
        if (largest != i) {
            int temp = arr[i];
            arr[i] = arr[largest];
            arr[largest] = temp;
            MaxHeapify(largest);
        }
    }

    int parent(int i) {
        return (i - 1) / 2;
    }

    int lChild(int i) {
        return (2 * i + 1);
    }

    int rChild(int i) {
        return (2 * i + 2);
    }

    int removeMax() {
        if (heapSize <= 0)
            return Integer.MIN_VALUE;
        if (heapSize == 1) {
            heapSize--;
            return arr[0];
        }
        int root = arr[0];
        arr[0] = arr[heapSize - 1];
        heapSize--;
        MaxHeapify(0);
        return root;
    }

    void increaseKey(int i, int newVal) {
        arr[i] = newVal;
        while (i != 0 && arr[parent(i)] < arr[i]) {
            int temp = arr[i];
            arr[i] = arr[parent(i)];
            arr[parent(i)] = temp;
            i = parent(i);
        }
    }

    int getMax() {
        return arr[0];
    }

    int curSize() {
        return heapSize;
    }

    void deleteKey(int i) {
        increaseKey(i, Integer.MAX_VALUE);
        removeMax();
    }

    void insertKey(int x) {
        if (heapSize == maxSize) {
            System.out.println("\nOverflow: Could not insertKey\n");
            return;
        }
        heapSize++;
        int i = heapSize - 1;
        arr[i] = x;
        while (i != 0 && arr[parent(i)] < arr[i]) {
            int temp = arr[i];
            arr[i] = arr[parent(i)];
            arr[parent(i)] = temp;
            i = parent(i);
        }
    }

    void printHeapArray() {
        for (int i = 0; i < heapSize; i++) {
            System.out.print(arr[i] + " ");
        }
        System.out.println();
    }

    void printHeapTree() {
        if (heapSize > 0) {
            System.out.println("Root: " + arr[0]);
            printHeapChildren(0);
        }
    }

    void printHeapChildren(int i) {
        int l = lChild(i);
        int r = rChild(i);
        if (l < heapSize) {
            System.out.println("LChild of " + arr[i] + ": " + arr[l]);
            printHeapChildren(l);
        }
        if (r < heapSize) {
            System.out.println("RChild of " + arr[i] + ": " + arr[r]);
            printHeapChildren(r);
        }
    }

    public static void main(String[] args) {
        MaxHeap heap = new MaxHeap(15);
        Scanner scanner = new Scanner(System.in);
        int choice;

        System.out.println("Mufid Bahaudin Nugroho");
        System.out.println("22106050021 / Informatika B");

        do {
            System.out.println("\nMENU:");
            System.out.println("1. Insert Element");
            System.out.println("2. Remove Max Element");
            System.out.println("3. Delete an Element at Index");
            System.out.println("4. Print MaxHeap V.1 (print array)");
            System.out.println("5. Print MaxHeap V.2 (print tree)");
            System.out.println("6. Quit");
            System.out.print("Enter your choice: ");
            choice = scanner.nextInt();

            switch (choice) {
                case 1:
                    System.out.print("Inserted Element:? ");
                    int element = scanner.nextInt();
                    heap.insertKey(element);
                    break;
                case 2:
                    System.out.println("Removed Max Element: " + heap.removeMax());
                    break;
                case 3:
                    System.out.print("Delete Element at Index:? ");
                    int index = scanner.nextInt();
                    heap.deleteKey(index);
                    break;
                case 4:
                    System.out.print("Heap Array: ");
                    heap.printHeapArray();
                    break;
                case 5:
                    System.out.println("Heap Tree:");
                    heap.printHeapTree();
                    break;
                case 6:
                    System.out.println("Exiting...");
                    break;
                default:
                    System.out.println("Pemilihan Invalid, silakan coba lagi!.");
                    break;
            }
        } while (choice != 6);

        scanner.close();
    }
}

File: Max Heap/maxheap.py
print("Nama : Mufid Bahaudin Nugroho")
print("22106050021")
print("Informatika B")
print("====================================")

class MaxHeap:
    def __init__(self, maxSize):
        self.maxSize = maxSize
        self.arr = [None] * maxSize
        self.heapSize = 0

    def MaxHeapify(self, i):
        l = self.lChild(i)
        r = self.rChild(i)
        largest = i
        if l < self.heapSize and self.arr[l] > self.arr[i]:
            largest = l
        if r < self.heapSize and self.arr[r] > self.arr[largest]:
            largest = r
        if largest != i:
            self.arr[i], self.arr[largest] = self.arr[largest], self.arr[i]
            self.MaxHeapify(largest)

    def parent(self, i):
        return (i - 1) // 2

    def lChild(self, i):
        return (2 * i + 1)

    def rChild(self, i):
        return (2 * i + 2)

    def removeMax(self):
        if self.heapSize <= 0:
            return None
        if self.heapSize == 1:
            self.heapSize -= 1
            return self.arr[0]
        root = self.arr[0]
        self.arr[0] = self.arr[self.heapSize - 1]
        self.heapSize -= 1
        self.MaxHeapify(0)
        return root

    def increaseKey(self, i, newVal):
        self.arr[i] = newVal
        while i != 0 and self.arr[self.parent(i)] < self.arr[i]:
            self.arr[i], self.arr[self.parent(i)] = self.arr[self.parent(i)], self.arr[i]
            i = self.parent(i)

    def getMax(self):
        return self.arr[0]

    def curSize(self):
        return self.heapSize

    def deleteKey(self, i):
        self.increaseKey(i, float("inf"))
        self.removeMax()

    def insertKey(self, x):
        if self.heapSize == self.maxSize:
            print("\nOverflow: Could not insertKey\n")
            return
        self.heapSize += 1
        i = self.heapSize - 1
        self.arr[i] = x
        while i != 0 and self.arr[self.parent(i)] < self.arr[i]:
            self.arr[i], self.arr[self.parent(i)] = self.arr[self.parent(i)], self.arr[i]
            i = self.parent(i)

    def printHeapArray(self):
        print("-".join(str(self.arr[i]) for i in range(self.heapSize)))

    def printHeapTree(self, i=0, depth=0):
        if i < self.heapSize:
            if i == 0:
                print(f"Root: {self.arr[i]}")
            l = self.lChild(i)
            r = self.rChild(i)
            if l < self.heapSize:
                print(f"LChild of {self.arr[i]}: {self.arr[l]}")
            if r < self.heapSize:
                print(f"RChild of {self.arr[i]}: {self.arr[r]}")
            if l < self.heapSize:
                self.printHeapTree(l, depth + 1)
            if r < self.heapSize:
                self.printHeapTree(r, depth + 1)


if __name__ == '__main__':
    h = MaxHeap(15)
    while True:
        print("\nMENU:")
        print("1. Insert Element")
        print("2. Remove Max Element")
        print("3. Delete an Element at Index")
        print("4. Print MaxHeap V.1 (print array)")
        print("5. Print MaxHeap V.2")
        print("6. Quit")
        choice = int(input("Enter your choice: "))

        if choice == 1:
            element = int(input("Inserted Element:? "))
            h.insertKey(element)
        elif choice == 2:
            print("Removed Max Element: " + str(h.removeMax()))
        elif choice == 3:
            index = int(input("Delete Element at Index:? "))
            h.deleteKey(index)
        elif choice == 4:
            print("Heap Array: ", end="")
            h.printHeapArray()
        elif choice == 5:
            print("Heap Tree:")
            h.printHeapTree()
        elif choice == 6:
            print("Exiting...")
            break
        else:
            print("Invalid choice. Please try again.")
